light-prover/prover/batch_update_circuit.go

```diff
@@ -0,0 +1,256 @@
+package prover
+
+import (
+	"fmt"
+	merkle_tree "light/light-prover/merkle-tree"
+	"light/light-prover/prover/poseidon"
+	"math/big"
+
+	"github.com/consensys/gnark-crypto/ecc"
+	"github.com/consensys/gnark/backend/groth16"
+	"github.com/consensys/gnark/constraint"
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/frontend/cs/r1cs"
+	"github.com/reilabs/gnark-lean-extractor/v2/abstractor"
+)
+
+type BatchUpdateCircuit struct {
+	OldRoot             frontend.Variable `gnark:",public"`
+	NewRoot             frontend.Variable `gnark:",public"`
+	HashChainStartIndex frontend.Variable `gnark:",public"`
```

Comment:
you can remove this I think it doesn't make sense afterall.
Instead we will require the number of batched zkp updates it requires to insert a batch into the tree to be fixed at tree initialization and create a separate LeavesHashchainHash for each zkp batch.